internal/wasi/wasi.go

```diff
@@ -1231,7 +1231,11 @@ func (a *wasiAPI) FdRead(ctx wasm.Module, fd, iovs, iovsCount, resultSize uint32
 
 	switch fd {
 	case 0:
-		reader = cfg.stdin
+		if cfg.stdin != nil {
+			reader = cfg.stdin
+		} else {
+			return wasi.ErrnoBadf
```

Comment:
I would prefer having Nil Reader rather than returning Badf as maybe all programs would expect `0` is valid FD in anycase

Reply:
ok just we have to close the reader, so I think I'll make a TODO as the code is not organized well enough now to close resources.

Reply:
unless by Nil reader you mean something besides os.DevNull...

Reply:
maybe passing `bytes.NewReader(nil)` is one idea :D

Reply:
remember this is for multiple modules (when many import one wasi), hence the close resources problem... and also if the program really needs to read stdin shouldn't it know somehow to override?

one way out is to TODO for a few days until there's proper resource management of file descriptors.

Reply:
Oh we need closer..

Reply:
yep up to you!

Reply:
the general problem in the current code is that file descriptors are wildly scoped, which will be solved at or after #394 as file descriptors should never be runtime scoped (ex they should be like memory, owned by the importing module)

Reply:
I'll make a code comment TODO as I've tried and it is too hard to get resources to close until we use standard types. If someone complains in any case they should be setting stdin if they actually need it, and badf error is not a horrible way to inform that for a few days.

thx for the nag it is a real issue I agree.

Reply:
sg!

Reply:
I have a better idea.. just took a 🚿 to think of it. iostreams aren't tightly coupled with memfs in the current code, so I can move them to module ahead of the preopens stuff. That has the advantage of lowering future scoped work, too. So, I'll update this in a bit!

Reply:
I found out that reading from os.DevNull was a very bad idea as it can run the host out of file descriptors easily if you forget to close. Reading EOF is a better default as if some code fails due to this, they should override stdin anyway